eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala

```diff
@@ -114,35 +648,407 @@ case class Commitments(channelId: ByteVector32,
+      active.foreach(_.canSendFee(cmd.feeratePerKw, params, changes1, feeConf) match {
+        case Some(f) => return Left(f)
+        case None => // ok
+      })
+      Right(copy(changes = changes1), fee)
+    }
+  }
+
+  def receiveFee(fee: UpdateFee, feeConf: OnChainFeeConf)(implicit log: LoggingAdapter): Either[ChannelException, Commitments] = {
+    if (params.localParams.isInitiator) {
+      Left(NonInitiatorCannotSendUpdateFee(channelId))
+    } else if (fee.feeratePerKw < FeeratePerKw.MinimumFeeratePerKw) {
+      Left(FeerateTooSmall(channelId, remoteFeeratePerKw = fee.feeratePerKw))
+    } else {
+      Metrics.RemoteFeeratePerKw.withoutTags().record(fee.feeratePerKw.toLong)
+      val localFeeratePerKw = feeConf.getCommitmentFeerate(params.remoteNodeId, params.channelType, active.head.capacity, None)
+      log.info("remote feeratePerKw={}, local feeratePerKw={}, ratio={}", fee.feeratePerKw, localFeeratePerKw, fee.feeratePerKw.toLong.toDouble / localFeeratePerKw.toLong)
+      // update_fee replace each other, so we can remove previous ones
+      val changes1 = changes.copy(remoteChanges = changes.remoteChanges.copy(proposed = changes.remoteChanges.proposed.filterNot(_.isInstanceOf[UpdateFee]) :+ fee))
+      active.foreach(_.canReceiveFee(fee.feeratePerKw, params, changes1, feeConf) match {
+        case Some(f) => return Left(f)
+        case None => // ok
+      })
+      Right(copy(changes = changes1))
+    }
+  }
 
+  def sendCommit(keyManager: ChannelKeyManager)(implicit log: LoggingAdapter): Either[ChannelException, (Commitments, Seq[CommitSig])] = {
+    remoteNextCommitInfo match {
+      case Right(_) if !changes.localHasChanges => Left(CannotSignWithoutChanges(channelId))
+      case Right(remoteNextPerCommitmentPoint) =>
+        val (active1, sigs) = active.map(_.sendCommit(keyManager, params, changes, remoteNextPerCommitmentPoint)).unzip
+        val commitments1 = copy(
+          changes = changes.copy(
+            localChanges = changes.localChanges.copy(proposed = Nil, signed = changes.localChanges.proposed),
+            remoteChanges = changes.remoteChanges.copy(acked = Nil, signed = changes.remoteChanges.acked),
+          ),
+          active = active1,
+          remoteNextCommitInfo = Left(WaitForRev(localCommitIndex))
+        )
+        Right(commitments1, sigs)
+      case Left(_) => Left(CannotSignBeforeRevocation(channelId))
+    }
+  }
+
+  def receiveCommit(commits: Seq[CommitSig], keyManager: ChannelKeyManager)(implicit log: LoggingAdapter): Either[ChannelException, (Commitments, RevokeAndAck)] = {
+    val channelKeyPath = keyManager.keyPath(params.localParams, params.channelConfig)
+    val localPerCommitmentPoint = keyManager.commitmentPoint(channelKeyPath, localCommitIndex + 1)
+    val active1 = active.map(_.receiveCommit(keyManager, params, changes, localPerCommitmentPoint, commits) match {
```

Comment:
It would make more sense that:
- `Commitments.receiveCommit` takes N sigs
- `Commitment.receiveCommit` takes 1 sig

The attribution of each sig to the correct commitment should be done in `Commitments.receiveCommit`, and a specific error should be raised if we don't have a perfect match. This would also greatly facilitate debugging compared to sending "invalid sigs".

Reply:
Implemented this and the actual aggregation in b9945d5c6. Feel free to cherry-pick 😄

Reply:
Good idea!

Reply:
Done in https://github.com/ACINQ/eclair/pull/2599/commits/4d00574833d77b225e85c2f52928a35bc9f49495, I removed the ordering requirement (not sure it's necessary) and added another error that specifies with commitment is missing from the set of `commit_sig`s, I believe this will be quite useful when debugging early splice prototypes!

Reply:
The ordering requirement allows us to simply not use the `fundingTxid`. I started with matching the `fundingTxid`, and indeed had to introduce an equivalent to your `CommitSigMissing`, but the code was tedious and the level of detail not very useful I think. I believe we may have obvious bugs at the beginning with mistmatch in active commitments, but they will quickly be ironed out. In the same way, we could provide the commitment number in `commit_sig`, right?

Reply:
That sounds good, I'll remove it and use implicit ordering for now instead, we can always add tlvs later if it becomes necessary

Reply:
Done in https://github.com/ACINQ/eclair/pull/2599/commits/3344632abcbda725acd3b7300e2a06c6b24b8c83